src/audio/dcblock/dcblock.c

```diff
@@ -0,0 +1,530 @@
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/dcblock/dcblock.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static void dcblock_free_config(struct sof_dcblock_config **config)
+{
+	rfree(*config);
+	*config = NULL;
+}
+
+static int dcblock_set_coeffs(struct comp_data *cd)
+{
+	int i;
+
+	/* Verify that the configuration values are set */
+	if (!cd->config)
+		return -EINVAL;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		cd->dcblock[i].R_coeff = cd->config->R_coeffs[i];
+
+	return 0;
+}
+
+/*
+ * Sets the DC Blocking filter in pass through mode.
+ * The frequency response of a DCB filter is:
+ * H(z) = (1 - z^-1)/(1-Rz^-1).
+ * Setting R to 1 makes the filter acts as a passthrough component.
+ */
+static int dcblock_set_passthrough(struct comp_data *cd)
+{
+	comp_cl_info(&comp_dcblock, "dcblock_set_passthrough()");
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		cd->dcblock[i].R_coeff = Q_CONVERT_FLOAT(1, 31);
+
+	return 0;
+}
+
+static int dcblock_set_state(struct comp_data *cd,
+			     struct audio_stream *src_stream)
+{
+	int32_t *x0;
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++) {
+		x0 = audio_stream_read_frag_s32(src_stream, 0);
+		cd->dcblock[i].y_prev = 0;
+
+		switch (src_stream->frame_fmt) {
+		case SOF_IPC_FRAME_S16_LE:
+			cd->dcblock[i].x_prev = *x0 << 16;
+			break;
+
+		case SOF_IPC_FRAME_S24_4LE:
+			cd->dcblock[i].x_prev = *x0 << 8;
+			break;
+
+		case SOF_IPC_FRAME_S32_LE:
+			cd->dcblock[i].x_prev = *x0;
+			break;
+
+		default:
+			comp_cl_err(&comp_dcblock, "dcblock_set_state(), "
+					"Invalid frame format: %i",
+					src_stream->frame_fmt);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * \brief Creates DC Blocking Filter component.
+ * \return Pointer to DC Blocking Filter component device.
+ */
+static struct comp_dev *dcblock_new(const struct comp_driver *drv,
+				    struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct comp_data *cd;
+	struct sof_ipc_comp_process *dcblock;
+	struct sof_ipc_comp_process *ipc_dcblock =
+		(struct sof_ipc_comp_process *)comp;
+	size_t bs = ipc_dcblock->size;
+	int ret;
+
+	comp_cl_info(&comp_dcblock, "dcblock_new()");
+
+	/* The blob size is fixed. Verify that it matches the expected
+	 * value before proceeding
+	 */
+	if (bs == sizeof(struct sof_dcblock_config)) {
```

Comment:
we should set ourselves to a passthrough if no config is passed from the topology. Tuning may come at runtime.